contracts/transfer-data/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+
+mod state;
+mod tree;
+
+use rusk_abi::ContractId;
+use state::TransferState;
+
+#[no_mangle]
+static SELF_ID: ContractId = ContractId::uninitialized();
+
+static mut STATE: TransferState = TransferState::new();
+
+// Queries
+
+#[no_mangle]
+unsafe fn root(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |_: ()| STATE.root())
+}
+
+#[no_mangle]
+unsafe fn module_balance(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |module| STATE.balance(&module))
+}
+
+#[no_mangle]
+unsafe fn message(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |(module, pk)| STATE.message(&module, &pk))
+}
+
+#[no_mangle]
+unsafe fn opening(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |pos| STATE.opening(pos))
+}
+
+#[no_mangle]
+unsafe fn existing_nullifiers(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |nullifiers| {
+        STATE.existing_nullifiers(nullifiers)
+    })
+}
+
+#[no_mangle]
+unsafe fn any_nullifier_exists(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |nullifiers| {
+        STATE.any_nullifier_exists(nullifiers)
+    })
+}
+
+#[no_mangle]
+unsafe fn extend_nullifiers(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |nullifiers| {
+        STATE.extend_nullifiers(nullifiers)
+    })
+}
+
+#[no_mangle]
+unsafe fn take_message_from_address_key(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |(module, pk)| {
+        STATE.take_message_from_address_key(&module, &pk)
+    })
+}
+
+#[no_mangle]
+unsafe fn root_exists(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |root| STATE.root_exists(&root))
+}
+
+#[no_mangle]
+unsafe fn push_message(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |(module, stealth_address, message)| {
+        STATE.push_message(module, stealth_address, message)
+    })
+}
+
+#[no_mangle]
+unsafe fn take_crossover(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |_: ()| STATE.take_crossover())
+}
+
+#[no_mangle]
+unsafe fn set_crossover(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |(crossover, address)| {
+        STATE.set_crossover(crossover, address)
+    })
+}
+
+#[no_mangle]
+unsafe fn get_crossover(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |_: ()| STATE.get_crossover())
+}
+
+#[no_mangle]
+unsafe fn extend_notes(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |(height, notes)| {
+        STATE.extend_notes(height, notes)
+    })
+}
+
+#[no_mangle]
+unsafe fn sub_balance(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |(module, value)| {
+        STATE.sub_balance(module, value)
+    })
+}
+
+// "Feeder" queries
+
+#[no_mangle]
+unsafe fn leaves_from_height(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |height| STATE.leaves_from_height(height))
+}
+
+#[no_mangle]
+unsafe fn leaves_from_pos(arg_len: u32) -> u32 {
+    rusk_abi::wrap_call(arg_len, |pos| STATE.leaves_from_pos(pos))
+}
+
+// "Management" transactions
+
+#[no_mangle]
+unsafe fn push_note(arg_len: u32) -> u32 {
```

Comment:
This (as all other management txs) should be called by transfer contract only

Reply:
And the above comments means that the caller cannot change due to upgradability (unless we implement some reserved contractId range)

Reply:
I think we should implement reserved contract id range in this case

Reply:
Thinking about it again... IMO all transfer-data methods should be called by the transfer-contract only.
We cannot open any of those calls to any other contract